src/sentry/types/region.py

```diff
@@ -119,42 +120,145 @@ class RegionContextError(Exception):
+                raise RegionConfigurationError(
+                    self._error_msg_for_region_setting("SENTRY_REGION", settings.SENTRY_REGION)
+                )
+
+    def _error_msg_for_region_setting(self, name: str, value: str) -> str:
+        return (
+            f"The {name} setting (value={value!r}) must point to a region name "
+            f"(region names = {[r.name for r in self.regions]!r})"
+        )
 
-        self.regions = frozenset(regions)
-        self.by_name = {r.name: r for r in self.regions}
+    def get(self, region_name: str) -> Region | None:
+        return self._by_name.get(region_name)
 
     def validate_all(self) -> None:
         for region in self.regions:
             region.validate()
 
 
-def parse_raw_config(region_config: Any) -> Iterable[Region]:
+class GlobalRegionDirectory:
+    """The set of all regions in this Sentry platform instance.
+
+    This class is a singleton. In a production environment, it is immutable.
+    Test environments support swapping its values in a temporary context.
+    """
+
+    def __init__(self, directory: RegionDirectory) -> None:
+        self._dir = directory
+        self._temporary_local_region: Region | None = None  # Used only in test env
+
+    @property
+    def regions(self) -> frozenset[Region]:
+        return self._dir.regions
+
+    @property
+    def historic_monolith_region(self) -> Region:
+        return self._dir.monolith_region
+
+    @property
+    def local_region(self) -> Region | None:
+        if self._temporary_local_region is not None:
+            self._allow_only_in_test_env()
+            return self._temporary_local_region
+        return self._dir.local_region
+
+    def get(self, region_name: str) -> Region | None:
+        return self._dir.get(region_name)
+
+    def get_all_region_names(self) -> Iterable[str]:
+        return (r.name for r in self.regions)
+
+    def _allow_only_in_test_env(self) -> None:
+        if not in_test_environment():
+            raise Exception("Swapping region values is allowed only in the test environment")
+
+    @contextmanager
+    def swap_directory(self, directory: RegionDirectory) -> Generator[None, None, None]:
+        self._allow_only_in_test_env()
+        old_dir = self._dir
+        try:
+            self._dir = directory
+            yield
+        finally:
+            self._dir = old_dir
+
+    @contextmanager
+    def swap_local_region(self, region: Region) -> Generator[None, None, None]:
+        self._allow_only_in_test_env()
+        if region not in self.regions:
+            raise Exception(
+                f"The swapped region {region.name} is not in this directory "
+                f"({[r.name for r in self.regions]})"
+            )
+
+        old_tmp_local_region = self._temporary_local_region
+        try:
+            self._temporary_local_region = region
+            yield
+        finally:
+            self._temporary_local_region = old_tmp_local_region
```

Comment:
This is an interview-question-level case of "not gonna work with parallel threads". That should be fine for unit tests.

I'm wondering how to square that with [the comment below](https://github.com/getsentry/sentry/blob/8b683002ec84f9edd338348500937a480e49e61c/src/sentry/types/region.py#L255-L260) regarding `single_process_silo_mode_state` and threaded acceptance tests. If we're lucky, I think we might actually be able to replace that with this, maybe with a simple mutex added. (The note about "we can't rely on settings" gives me hope, because the whole point here is not to look at Django settings in the first place.) @corps?